packages/signers/src/lit-protocol/signer.ts

```diff
@@ -0,0 +1,196 @@
+import { PKPEthersWallet } from "@lit-protocol/pkp-ethers";
+import { LitAbility, LitActionResource } from "@lit-protocol/auth-helpers";
+import {
+  type AuthCallbackParams,
+  type AuthSig,
+  type SessionSigsMap,
+} from "@lit-protocol/types";
+import { LitNodeClient } from "@lit-protocol/lit-node-client";
+import { generateSessionKeyPair } from "@lit-protocol/crypto";
+
+import { type SignTypedDataParams } from "@alchemy/aa-core";
+import type { Address, TypedDataDomain } from "viem";
+import {
+  type LitAuthMethod,
+  type LitSessionSigsMap,
+  type LitConfig,
+  type LitAuthenticateProps,
+  type LitSmartAccountAuthenticator,
+  type LitUserMetadata,
+} from "./types.js";
+
+const SIGNER_TYPE: string = "lit";
+
+/**
+ * Implementation of `SmartAccountAuthenticator` for lit protocol
+ * This class requies:
+ * `@lit-protocol/lit-node-client@cayenne`
+ * `@lit-protocol/pkp-ethers@cayenne`
+ * `@lit-protocol/crypto@cayenne`
+ * `@lit-protocol/auth-helpers@cayenne`
+ * `@lit-protocol/types@cayenne`
+ */
+export class LitSigner<C extends LitAuthMethod | LitSessionSigsMap>
+  implements LitSmartAccountAuthenticator<C>
+{
+  inner: LitNodeClient;
+  public signer: PKPEthersWallet | undefined;
+  private _pkpPublicKey: string;
+  private _rpcUrl: string;
+  private _authContext: C | undefined;
+  private _session: SessionSigsMap | undefined;
+
+  constructor(params: LitConfig) {
+    this._pkpPublicKey = params.pkpPublicKey;
+    this.inner =
+      params.inner ??
+      new LitNodeClient({
+        litNetwork: params.network ?? "cayenne",
+        debug: params.debug ?? false,
+      });
+    this._rpcUrl = params.rpcUrl;
+  }
+  signerType: string = SIGNER_TYPE;
+
+  /**
+   * if generic type is `LitAuthMethod`, authenticates the supplied authentication material.
+   * if type `SessionSigsMap`, this implementation will respect the existing auth and use the session material.
+   * @param props {LITAuthenticateProps} Authentication params, only `context` is required
+   * @returns {LitSessionSigsMap} Authenticated session material
+   */
+  authenticate = async (
```

Comment:
I'm not too familiar with how Lit manages persistent sessions for users, but is that supported here? or does authenticate have to be called with input from a user each time?

my point of reference is the MagicSDK. The SDK maintains session state so it's possible to be authenticated without calling `authenticate`

Reply:
Our sdk does not hold the authentication state, it is required to be passed by the caller. We could check if the authentication material is already present in the state of `LitSigner` and not perform the logic again for now. once we have a internal caching layer can refactor to no longer need the implementation shown here:

For now ive moved the logic of getting a session to `doAuthentication` and a check on `session` not being defined to run the authentication.